src/kaleidoscope/plugin/OneShot.cpp

```diff
@@ -22,266 +22,339 @@
 namespace kaleidoscope {
 namespace plugin {
 
-// ---- state ---------
+// ----------------------------------------------------------------------------
+// Configuration variables
 
-uint16_t OneShot::start_time_ = 0;
 uint16_t OneShot::time_out = 2500;
 uint16_t OneShot::hold_time_out = 250;
 int16_t OneShot::double_tap_time_out = -1;
-OneShot::key_state_t OneShot::state_[OneShot::ONESHOT_KEY_COUNT];
-Key OneShot::prev_key_;
-bool OneShot::should_cancel_ = false;
-bool OneShot::should_cancel_stickies_ = false;
 
-bool OneShot::isPressed() {
-  for (uint8_t i = 0; i < ONESHOT_KEY_COUNT; i++) {
-    if (state_[i].pressed)
+// ----------------------------------------------------------------------------
+// State variables
+
+OneShot::OneShotKeyState OneShot::state_[OneShot::ONESHOT_KEY_COUNT];
+uint16_t OneShot::start_time_ = 0;
+KeyAddr OneShot::prev_key_addr_ = UnknownKeyswitchLocation;
+uint8_t OneShot::release_countdown_ = 0;
+
+
+// ============================================================================
+// Public interface
+
+// ----------------------------------------------------------------------------
+// Configuration functions
+
+void OneShot::enableStickability(Key key) {
+  if (key >= ranges::OS_FIRST && key <= ranges::OS_LAST)
+    state_[key.getRaw() - ranges::OS_FIRST].stickable = true;
+}
+
+void OneShot::disableStickability(Key key) {
+  if (key >= ranges::OS_FIRST && key <= ranges::OS_LAST)
+    state_[key.getRaw() - ranges::OS_FIRST].stickable = false;
+}
+
+void OneShot::enableStickabilityForModifiers() {
+  for (uint8_t i{0}; i < ONESHOT_MOD_COUNT; ++i) {
+    state_[i].stickable = true;
+  }
+}
+
+void OneShot::disableStickabilityForModifiers() {
+  for (uint8_t i{0}; i < ONESHOT_MOD_COUNT; ++i) {
+    state_[i].stickable = false;
+  }
+}
+
+void OneShot::enableStickabilityForLayers() {
+  for (uint8_t i{ONESHOT_MOD_COUNT}; i < ONESHOT_KEY_COUNT; ++i) {
```

Comment:
The existing version of this function (and the one below) used `ONESHOT_KEY_COUNT / 2`, taking advantage of what is essentially a coincidence (there happen to be the same number of modifiers and supported OSL values). I think this illustrates more clearly what's going on, and why.